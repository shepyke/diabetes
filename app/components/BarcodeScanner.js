/**
 * Diabetes
 * https://github.com/shepyke/diabetes
 * @flow
 */

import React, { Component } from 'react';
import {
    Alert,
    View
} from 'react-native';
import {RNCamera} from 'react-native-camera';
import Styles from "../config/Styles";
import AddFood from "./AddFood";

export default class BarcodeScanner extends Component {
    constructor(props) {
        super(props);

        this.state = {
            camera: {
                type: RNCamera.Constants.Type.back,
                flashMode: RNCamera.Constants.FlashMode.auto,
            },
            showCamera: true,
            foods: [
                {
                    foodId: '',
                    barcode: '',
                    calorie: '',
                    carbohydrate: '',
                    fat: '',
                    foodName: '',
                    glycemicIndex: '',
                    photo: '',
                    protein:'',
                    unit: '',
                },
            ],
            food: {
                foodId: '',
                barcode: '',
                calorie: '',
                carbohydrate: '',
                fat: '',
                foodName: '',
                glycemicIndex: '',
                photo: '',
                protein:'',
                unit: '',
            },
            intakeId: (this.props.navigation.state.params != undefined) ? this.props.navigation.state.params.id : -1,
            viaIntake: (this.props.navigation.state.params != undefined) ? this.props.navigation.state.params.viaIntake : false,
        };
    }

    componentDidMount(){
        this._loadInitialState().done();
    }

    componentWillReceiveProps(nextProps){
        if(nextProps.navigation.state.params != undefined){
            if(nextProps.navigation.state.params.viaIntake != undefined) {
                this.setState({
                    intakeId: nextProps.navigation.state.params.id,
                    viaIntake: nextProps.navigation.state.params.viaIntake,
                });
            }
        }
    }

    _loadInitialState = async() => {
        try {
            fetch('https://diabetes-backend.herokuapp.com/intakes/foods')
                .then((response) => response.json())
                .then((res) => {
                    this.state.foods = res.foods;
                })
                .catch((error) => {
                    console.error(error);
                });
        }catch(error){
            console.error(error);
        }

        this._sub = this.props.navigation.addListener('didFocus', () => {
            this.setState({
                showCamera: true,
            });
        });
    }

    componentWillUnmount() {
        this._sub.remove();
    }

    findBarcode(foodsObject, barcode) {
        return foodsObject.find((element) => {
            if(element.barcode == barcode){
                return element;
            }else{
                return undefined;
            }
        })
    }

    onBarCodeRead(barcode) {
        let food = this.findBarcode(this.state.foods, barcode.data);
        if(food != undefined){
            this.setState({
                showCamera: false,
            });
            if(this.state.viaIntake) {
                this.props.navigation.navigate('Intake', {foodId: food.foodId, newElement: false, intakeId: this.state.intakeId});
            }else{
                this.props.navigation.navigate('Intake', {foodId: food.foodId, newElement: true})
            }
            this.setState({
                viaIntake: false,
            });
        }else{
            this.setState({
                showCamera: false,
            });
            Alert.alert(
                'Sorry',
                'not found any food with this barcode',
                [
                    {text: 'Ok', onPress: () => {
                            console.log('Cancel Pressed');
                            this.setState({
                               showCamera: true,
                            });
                    }, style: 'cancel'},
                    {text: 'Add as a new food', onPress:() => {
                            this.refs.addNewFood.showAddFoodModal(barcode);
                        }
                    },
                ],
                { cancelable: false }
            )
        }
    }

    render() {
        if(this.state.showCamera) {
            return (
                <View style={Styles.wrapper}>
                    <RNCamera
                        ref={(cam) => {
                            this.camera = cam;
                        }}
                        onBarCodeRead={this.onBarCodeRead.bind(this)}
                        style={Styles.preview}
                        type={this.state.camera.type}
                        torchMode={this.state.camera.flashMode}
                        defaultTouchToFocus
                        mirrorImage={false}
                    />
                    <AddFood ref={'addNewFood'}/>
                </View>
            );
        }else{
            return (
                <View style={Styles.wrapper}>
                    <AddFood ref={'addNewFood'}/>
                </View>
            );
        }
    }

}