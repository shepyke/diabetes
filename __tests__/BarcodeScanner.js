import React from 'react';
import 'react-native';
//import BarcodeScanner from '../app/components/BarcodeScanner';

import renderer from 'react-test-renderer';

import BarcodeScanner from '../app/components/__mocks__/react-native-camera';

jest.doMock('react-native-camera', () => BarcodeScanner);


//Start of testId: barcode-1
it('Camera initialization is done', () => {
    const tree = renderer.create(<BarcodeScanner />).toJSON();
    expect(tree).toMatchSnapshot();
})
//End of testId: barcode-1

//Start of testId: barcode-2
test('the given barcode is part of the database', () => {
    const foods = [
        {"foodId":1,"glycemicIndex":1,"carbohydrate":45.8,"fat":6.5,"calorie":285,"protein":10.8,"photo":"","barcode":0,"foodName":"Magvas barnakenyér","regTime":"2018-04-20T11:03:45.000Z","unit":"g"},
        {"foodId":2,"glycemicIndex":2,"carbohydrate":33.3,"fat":33,"calorie":1,"protein":13,"photo":"","barcode":11,"foodName":"Virsli","regTime":"2018-04-20T11:03:45.000Z","unit":"g"},
        {"foodId":3,"glycemicIndex":2,"carbohydrate":4.7,"fat":2.8,"calorie":56,"protein":3,"photo":"","barcode":5995805038851,"foodName":"Finest Tej","regTime":"2018-04-20T11:03:45.000Z","unit":"ml"},
        {"foodId":4,"glycemicIndex":0,"carbohydrate":3.5,"fat":0,"calorie":43,"protein":0.4,"photo":"","barcode":5999601340156,"foodName":"Space Next - Rocket","regTime":"2018-04-20T11:03:45.000Z","unit":"ml"},
        {"foodId":5,"glycemicIndex":2,"carbohydrate":4.7,"fat":2.8,"calorie":56,"protein":3,"photo":"","barcode":5998202600034,"foodName":"Kunsági Tej","regTime":"2018-04-20T11:03:45.000Z","unit":"ml"},
        {"foodId":6,"glycemicIndex":2,"carbohydrate":2.2,"fat":0.1,"calorie":10,"protein":0.1,"photo":"","barcode":5998821500777,"foodName":"Olympos Light","regTime":"2018-04-20T11:03:45.000Z","unit":"ml"},
        {"foodId":7,"glycemicIndex":0,"carbohydrate":3.5,"fat":0,"calorie":43,"protein":0.4,"photo":"","barcode":8594404115115,"foodName":"Pilsner Urquell","regTime":"2018-04-20T11:03:45.000Z","unit":"ml"}
    ];

    expect(foods).toEqual(          // 1
        expect.arrayContaining([      // 2
            expect.objectContaining({   // 3
                "barcode": 5999601340156               // 4
            })
        ])
    );
})
//End of testId: barcode-2

//Start of testId: barcode-3
it('GetFoods is not empty', () => {
    const tree = renderer.create(<BarcodeScanner />).getInstance();
    expect(tree.getFoods).not.toBeNull();
})
//End of testId: barcode-3