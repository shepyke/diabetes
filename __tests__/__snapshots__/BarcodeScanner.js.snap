// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera initialization is done 1`] = `null`;
